Test failures with Django 3.2 alpha 1

Hello,

Thanks for your work on django-modeltranslation.

I wanted to see if django-modeltranslation was ready for Django 3.2 and it looks like this is not yet the case. I noticed the documentation (docs/modeltranslation/installation.rst) lacks support information about Python 3.9 and Django 3.1 as well.

With 3.2 alpha 1 I get those failures:

```
I: pybuild base:232: python3.9 ./runtests.py
/<<PKGBUILDDIR>>/modeltranslation/tests/tests.py:6: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Creating test database for alias 'default' ('file:memorydb_default?mode=memory&cache=shared')...
Operations to perform:
  Synchronize unmigrated apps: modeltranslation
  Apply all migrations: auth, contenttypes
Synchronizing apps without migrations:
  Creating tables...
    Running deferred SQL...
Running migrations:
  Applying contenttypes.0001_initial... OK
  Applying contenttypes.0002_remove_content_type_name... OK
  Applying auth.0001_initial... OK
  Applying auth.0002_alter_permission_name_max_length... OK
  Applying auth.0003_alter_user_email_max_length... OK
  Applying auth.0004_alter_user_username_opts... OK
  Applying auth.0005_alter_user_last_login_null... OK
  Applying auth.0006_require_contenttypes_0002... OK
  Applying auth.0007_alter_validators_add_error_messages... OK
  Applying auth.0008_alter_user_username_max_length... OK
  Applying auth.0009_alter_user_last_name_max_length... OK
  Applying auth.0010_alter_group_name_max_length... OK
  Applying auth.0011_update_proxy_permissions... OK
  Applying auth.0012_alter_user_first_name_max_length... OK
System check identified no issues (0 silenced).
Migrations for 'auth':
  modeltranslation/tests/auth_migrations/0013_group_name_en.py
    - Add field name_en to group
Migrations for 'auth':
  modeltranslation/tests/auth_migrations/0014_auto_20210122_2045.py
    - Change managers on group
    - Change managers on permission
    - Change managers on user
    - Add field name_de to group
test_fallback_languages (modeltranslation.tests.tests.FallbackTests) ... ok
test_fallback_undefined (modeltranslation.tests.tests.FallbackTests)
Checks if a sensible value is considered undefined and triggers ... ok
test_fallbacks_toggle (modeltranslation.tests.tests.FallbackTests) ... ok
test_resolution_order (modeltranslation.tests.tests.FallbackTests) ... ok
test_settings (modeltranslation.tests.tests.FallbackTests) ... ok
test_empty_field (modeltranslation.tests.tests.FileFieldsTest) ... ok
test_fallback (modeltranslation.tests.tests.FileFieldsTest) ... ok
test_translated_models (modeltranslation.tests.tests.FileFieldsTest) ... ok
test_translated_models_instance (modeltranslation.tests.tests.FileFieldsTest) ... ok
test_db_column_names (modeltranslation.tests.tests.ForeignKeyFieldsTest) ... ok
test_indonesian (modeltranslation.tests.tests.ForeignKeyFieldsTest) ... ok
test_non_translated_relation (modeltranslation.tests.tests.ForeignKeyFieldsTest) ... FAIL
test_reverse_lookup_with_filtered_queryset_manager (modeltranslation.tests.tests.ForeignKeyFieldsTest)
Make sure base_manager does not get same queryset filter as TestModel in reverse lookup ... ok
test_reverse_relations (modeltranslation.tests.tests.ForeignKeyFieldsTest) ... FAIL
test_translated_models (modeltranslation.tests.tests.ForeignKeyFieldsTest) ... ok
test_translated_models_instance (modeltranslation.tests.tests.ForeignKeyFieldsTest) ... FAIL
test_managers_failure (modeltranslation.tests.tests.InheritedPermissionTestCase)
This fails with 0.13b. ... ok
test_m2m (modeltranslation.tests.tests.M2MTest) ... FAIL
test_field_aggregation (modeltranslation.tests.tests.ModelInheritanceFieldAggregationTest) ... ok
test_multi_inheritance (modeltranslation.tests.tests.ModelInheritanceFieldAggregationTest) ... ok
test_abstract_inheritance (modeltranslation.tests.tests.ModelInheritanceTest) ... ok
test_inheritance (modeltranslation.tests.tests.ModelInheritanceTest) ... ok
test_multitable_inheritance (modeltranslation.tests.tests.ModelInheritanceTest) ... ok
test_model_validation_email_field (modeltranslation.tests.tests.ModelValidationTest) ... ok
test_model_validation_required (modeltranslation.tests.tests.ModelValidationTest)
General test for CharField: if required/blank is handled properly. ... ok
test_model_validation_url_field (modeltranslation.tests.tests.ModelValidationTest) ... ok
test_constructor (modeltranslation.tests.tests.ModeltranslationTest)
Ensure that model constructor behaves exactly the same as objects.create ... ok
test_descriptor_introspection (modeltranslation.tests.tests.ModeltranslationTest) ... ok
test_fallback_language (modeltranslation.tests.tests.ModeltranslationTest) ... ok
test_fallback_values_1 (modeltranslation.tests.tests.ModeltranslationTest)
If ``fallback_values`` is set to string, all untranslated fields would ... ok
test_fallback_values_2 (modeltranslation.tests.tests.ModeltranslationTest)
If ``fallback_values`` is set to ``dict``, all untranslated fields in ... ok
test_fields (modeltranslation.tests.tests.ModeltranslationTest) ... ok
test_fields_hashes (modeltranslation.tests.tests.ModeltranslationTest) ... ok
test_registration (modeltranslation.tests.tests.ModeltranslationTest) ... ok
test_registration_field_conflicts (modeltranslation.tests.tests.ModeltranslationTest) ... ok
test_set_translation (modeltranslation.tests.tests.ModeltranslationTest)
This test briefly shows main modeltranslation features. ... FAIL
test_verbose_name (modeltranslation.tests.tests.ModeltranslationTest) ... ok
test_rule1 (modeltranslation.tests.tests.ModeltranslationTestRule1)
Basic CharField/TextField test. ... ok
test_rule1_email_field (modeltranslation.tests.tests.ModeltranslationTestRule1) ... ok
test_rule1_url_field (modeltranslation.tests.tests.ModeltranslationTestRule1) ... ok
test_rule2 (modeltranslation.tests.tests.ModeltranslationTestRule2)
Basic CharField/TextField test. ... ok
test_rule2_email_field (modeltranslation.tests.tests.ModeltranslationTestRule2) ... ok
test_rule2_url_field (modeltranslation.tests.tests.ModeltranslationTestRule2) ... ok
test_rule3 (modeltranslation.tests.tests.ModeltranslationTestRule3) ... ok
test_rule3_internals (modeltranslation.tests.tests.ModeltranslationTestRule3) ... ok
test_db_column_names (modeltranslation.tests.tests.OneToOneFieldsTest) ... ok
test_indonesian (modeltranslation.tests.tests.OneToOneFieldsTest) ... ok
test_non_translated_relation (modeltranslation.tests.tests.OneToOneFieldsTest) ... FAIL
test_reverse_lookup_with_filtered_queryset_manager (modeltranslation.tests.tests.OneToOneFieldsTest)
Make sure base_manager does not get same queryset filter as TestModel in reverse lookup ... ok
test_reverse_relations (modeltranslation.tests.tests.OneToOneFieldsTest) ... FAIL
test_translated_models (modeltranslation.tests.tests.OneToOneFieldsTest) ... ok
test_translated_models_instance (modeltranslation.tests.tests.OneToOneFieldsTest) ... FAIL
test_uniqueness (modeltranslation.tests.tests.OneToOneFieldsTest) ... ok
test_dates_queryset (modeltranslation.tests.tests.OtherFieldsTest) ... ok
test_descriptors (modeltranslation.tests.tests.OtherFieldsTest) ... ok
test_translated_models (modeltranslation.tests.tests.OtherFieldsTest) ... ok
test_translated_models_boolean_instance (modeltranslation.tests.tests.OtherFieldsTest) ... ok
test_translated_models_commaseparatedinteger_instance (modeltranslation.tests.tests.OtherFieldsTest) ... ok
test_translated_models_date_instance (modeltranslation.tests.tests.OtherFieldsTest) ... ok
test_translated_models_datetime_instance (modeltranslation.tests.tests.OtherFieldsTest) ... ok
test_translated_models_decimal_instance (modeltranslation.tests.tests.OtherFieldsTest) ... ok
test_translated_models_float_instance (modeltranslation.tests.tests.OtherFieldsTest) ... ok
test_translated_models_genericipaddress_instance (modeltranslation.tests.tests.OtherFieldsTest) ... ok
test_translated_models_integer_instance (modeltranslation.tests.tests.OtherFieldsTest) ... ok
test_translated_models_ipaddress_instance (modeltranslation.tests.tests.OtherFieldsTest) ... ok
test_translated_models_nullboolean_instance (modeltranslation.tests.tests.OtherFieldsTest) ... ok
test_translated_models_time_instance (modeltranslation.tests.tests.OtherFieldsTest) ... ok
test_equality (modeltranslation.tests.tests.ProxyModelTest) ... ok
test_duplication (modeltranslation.tests.tests.TestAutodiscover)
Check if there is no problem with duplicated filenames. ... ok
test_global (modeltranslation.tests.tests.TestAutodiscover)
Check if translation is imported for global translation file. ... ok
test_simple (modeltranslation.tests.tests.TestAutodiscover)
Check if translation is imported for installed apps. ... ok
test_3rd_party_custom_manager (modeltranslation.tests.tests.TestManager) ... ok
test_constructor_inheritance (modeltranslation.tests.tests.TestManager) ... ok
test_creation (modeltranslation.tests.tests.TestManager)
Test if field are rewritten in create. ... ok
test_creation_population (modeltranslation.tests.tests.TestManager)
Test if language fields are populated with default value on creation. ... ok
test_custom_manager (modeltranslation.tests.tests.TestManager)
Test if user-defined manager is still working ... FAIL
test_custom_manager2 (modeltranslation.tests.tests.TestManager)
Test if user-defined queryset is still working ... ok
test_custom_manager_custom_method_name (modeltranslation.tests.tests.TestManager)
Test if custom method also returns MultilingualQuerySet ... ok
test_default_manager_for_inherited_models (modeltranslation.tests.tests.TestManager) ... ok
test_default_manager_for_inherited_models_with_custom_manager (modeltranslation.tests.tests.TestManager)
Test if default manager is still set from local managers ... ok
test_deferred (modeltranslation.tests.tests.TestManager)
Check if ``only`` and ``defer`` are working. ... ok
test_deferred_fk (modeltranslation.tests.tests.TestManager)
Check if ``select_related`` is rewritten and also ... ok
test_deferred_rule2 (modeltranslation.tests.tests.TestManager) ... ok
test_deferred_spanning (modeltranslation.tests.tests.TestManager) ... ok
test_distinct (modeltranslation.tests.tests.TestManager)
Check that field names are rewritten in distinct keys. ... ok
test_f (modeltranslation.tests.tests.TestManager)
Test if F queries are rewritten. ... ok
test_filter_update (modeltranslation.tests.tests.TestManager)
Test if filtering and updating is language-aware. ... FAIL
test_fixture_population (modeltranslation.tests.tests.TestManager)
Test that a fixture with values only for the original fields ... ok
test_fixture_population_via_command (modeltranslation.tests.tests.TestManager)
Test that the loaddata command takes new option. ... ok
test_get_or_create_population (modeltranslation.tests.tests.TestManager)
Populate may be used with ``get_or_create``. ... ok
test_multilingual_queryset_pickling (modeltranslation.tests.tests.TestManager) ... ok
test_non_objects_manager (modeltranslation.tests.tests.TestManager)
Test if managers other than ``objects`` are patched too ... ok
test_order_by (modeltranslation.tests.tests.TestManager)
Check that field names are rewritten in order_by keys. ... ok
test_order_by_meta (modeltranslation.tests.tests.TestManager)
Check that meta ordering is rewritten. ... ok
test_q (modeltranslation.tests.tests.TestManager)
Test if Q queries are rewritten. ... FAIL
test_select_related_django_1 (modeltranslation.tests.tests.TestManager) ... skipped 'Applicable only to django 1.x'
test_select_related_django_2 (modeltranslation.tests.tests.TestManager) ... ok
test_translation_fields_appending (modeltranslation.tests.tests.TestManager) ... ok
test_values (modeltranslation.tests.tests.TestManager) ... ok
test_values_fallback (modeltranslation.tests.tests.TestManager) ... ok
test_values_list_annotation (modeltranslation.tests.tests.TestManager) ... ok
test_required (modeltranslation.tests.tests.TestRequired) ... ok
test_form (modeltranslation.tests.tests.ThirdPartyAppIntegrationRegisteredTest) ... ok
test_form (modeltranslation.tests.tests.ThirdPartyAppIntegrationTest) ... ok
test_build_css_class (modeltranslation.tests.tests.TranslationAdminTest) ... ok
test_default_fields (modeltranslation.tests.tests.TranslationAdminTest) ... ok
test_default_fieldsets (modeltranslation.tests.tests.TranslationAdminTest) ... ok
test_field_arguments (modeltranslation.tests.tests.TranslationAdminTest) ... ok
test_field_arguments_restricted_on_custom_form (modeltranslation.tests.tests.TranslationAdminTest) ... ok
test_field_arguments_restricted_on_form (modeltranslation.tests.tests.TranslationAdminTest) ... ok
test_group_fieldsets (modeltranslation.tests.tests.TranslationAdminTest) ... ok
test_inline_fieldsets (modeltranslation.tests.tests.TranslationAdminTest) ... ok
test_list_editable (modeltranslation.tests.tests.TranslationAdminTest) ... ok
test_model_form_widgets (modeltranslation.tests.tests.TranslationAdminTest) ... ok
test_multitable_inheritance (modeltranslation.tests.tests.TranslationAdminTest) ... ok
test_prepopulated_fields (modeltranslation.tests.tests.TranslationAdminTest) ... ok
test_proxymodel_field_argument (modeltranslation.tests.tests.TranslationAdminTest) ... ok
test_widget_ordering_via_formfield_for_dbfield (modeltranslation.tests.tests.TranslationAdminTest) ... ok
test_fields (modeltranslation.tests.tests.TranslationModelFormTest) ... ok
test_updating_with_empty_value (modeltranslation.tests.tests.TranslationModelFormTest)
Can we update the current language translation with an empty value, when ... ok
test_update_command (modeltranslation.tests.tests.UpdateCommandTest) ... ok
test_update_command_invalid_language_param (modeltranslation.tests.tests.UpdateCommandTest) ... ok
test_update_command_language_param (modeltranslation.tests.tests.UpdateCommandTest) ... ok
test_unique_nullable_field (modeltranslation.tests.tests.ModeltranslationTransactionTest) ... ok

======================================================================
FAIL: test_non_translated_relation (modeltranslation.tests.tests.ForeignKeyFieldsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/modeltranslation/tests/tests.py", line 1047, in test_non_translated_relation
    self.assertEqual(manager.filter(non=non_en).count(), 2)
AssertionError: 0 != 2

======================================================================
FAIL: test_reverse_relations (modeltranslation.tests.tests.ForeignKeyFieldsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/modeltranslation/tests/tests.py", line 949, in test_reverse_relations
    self.assertIn(fk_inst_en,    test_inst.test_fks.all())
AssertionError: <ForeignKeyModel: ForeignKeyModel object (3)> not found in <MultilingualQuerySet [<ForeignKeyModel: ForeignKeyModel object (1)>, <ForeignKeyModel: ForeignKeyModel object (2)>]>

======================================================================
FAIL: test_translated_models_instance (modeltranslation.tests.tests.ForeignKeyFieldsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/modeltranslation/tests/tests.py", line 878, in test_translated_models_instance
    self.assertEqual(manager.filter(test=test_inst1).count(), 1)
AssertionError: 0 != 1

======================================================================
FAIL: test_m2m (modeltranslation.tests.tests.M2MTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/modeltranslation/tests/tests.py", line 3271, in test_m2m
    self.assertEqual(1, y_foo.count())
AssertionError: 1 != 0

======================================================================
FAIL: test_set_translation (modeltranslation.tests.tests.ModeltranslationTest)
This test briefly shows main modeltranslation features.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/modeltranslation/tests/tests.py", line 453, in test_set_translation
    self.assertEqual(0, models.TestModel.objects.filter(title=title_de).count())
AssertionError: 0 != 1

======================================================================
FAIL: test_non_translated_relation (modeltranslation.tests.tests.OneToOneFieldsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/modeltranslation/tests/tests.py", line 1220, in test_non_translated_relation
    self.assertEqual(manager.filter(non=non_en).count(), 1)
AssertionError: 0 != 1

======================================================================
FAIL: test_reverse_relations (modeltranslation.tests.tests.OneToOneFieldsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/modeltranslation/tests/tests.py", line 1171, in test_reverse_relations
    self.assertEqual(manager.filter(test_o2o=fk_inst_en).count(), 1)
AssertionError: 0 != 1

======================================================================
FAIL: test_translated_models_instance (modeltranslation.tests.tests.OneToOneFieldsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/modeltranslation/tests/tests.py", line 878, in test_translated_models_instance
    self.assertEqual(manager.filter(test=test_inst1).count(), 1)
AssertionError: 0 != 1

======================================================================
FAIL: test_custom_manager (modeltranslation.tests.tests.TestManager)
Test if user-defined manager is still working
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/modeltranslation/tests/tests.py", line 2772, in test_custom_manager
    self.assertEqual(1, models.CustomManagerTestModel.objects.count())
AssertionError: 1 != 2

======================================================================
FAIL: test_filter_update (modeltranslation.tests.tests.TestManager)
Test if filtering and updating is language-aware.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/modeltranslation/tests/tests.py", line 2563, in test_filter_update
    self.assertEqual(2, models.ManagerTestModel.objects.filter(title='de').count())
AssertionError: 2 != 0

======================================================================
FAIL: test_q (modeltranslation.tests.tests.TestManager)
Test if Q queries are rewritten.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/modeltranslation/tests/tests.py", line 2600, in test_q
    self.assertEqual(1, models.ManagerTestModel.objects.filter(Q(title='de') |
AssertionError: 1 != 0

----------------------------------------------------------------------
Ran 124 tests in 0.847s

FAILED (failures=11, skipped=1)
Destroying test database for alias 'default' ('file:memorydb_default?mode=memory&cache=shared')...

```

--- a/.travis.yml
+++ b/.travis.yml
@@ -17,6 +17,8 @@ env:
   - DJANGO=3.0 DB=sqlite
   - DJANGO=3.0 DB=postgres
   - DJANGO=3.0 DB=mysql
+  - DJANGO=3.2 DB=postgres
+  - DJANGO=3.2 DB=mysql
 before_install:
   - pip install -q flake8
   - PYFLAKES_NODOCTEST=1 flake8 modeltranslation

--- a/modeltranslation/management/commands/update_translation_fields.py
+++ b/modeltranslation/management/commands/update_translation_fields.py
@@ -81,6 +81,6 @@ class Command(BaseCommand):
                 if field.empty_strings_allowed:
                     q |= Q(**{def_lang_fieldname: ""})
 
-                model._default_manager.filter(q).rewrite(False).update(
+                model._default_manager.filter(q).rewrite(False).order_by().update(
                     **{def_lang_fieldname: F(field_name)}
                 )

--- a/modeltranslation/manager.py
+++ b/modeltranslation/manager.py
@@ -7,6 +7,7 @@ https://github.com/zmathew/django-linguo
 """
 import itertools
 
+from django import VERSION
 from django.contrib.admin.utils import get_model_from_relation
 from django.core.exceptions import FieldDoesNotExist
 from django.db import models
@@ -296,15 +297,24 @@ class MultilingualQuerySet(models.query.QuerySet):
             q.rhs = self._rewrite_f(q.rhs)
         return q
 
-    def _filter_or_exclude(self, negate, *args, **kwargs):
+    def _rewrite_filter_or_exclude(self, args, kwargs):
         if not self._rewrite:
-            return super(MultilingualQuerySet, self)._filter_or_exclude(negate, *args, **kwargs)
+            return args, kwargs
         args = map(self._rewrite_q, args)
         for key, val in list(kwargs.items()):
             new_key = rewrite_lookup_key(self.model, key)
             del kwargs[key]
             kwargs[new_key] = self._rewrite_f(val)
-        return super(MultilingualQuerySet, self)._filter_or_exclude(negate, *args, **kwargs)
+        return args, kwargs
+
+    if VERSION >= (3, 2):
+        def _filter_or_exclude(self, negate, args, kwargs):
+            args, kwargs = self._rewrite_filter_or_exclude(args, kwargs)
+            return super(MultilingualQuerySet, self)._filter_or_exclude(negate, args, kwargs)
+    else:
+        def _filter_or_exclude(self, negate, *args, **kwargs):
+            args, kwargs = self._rewrite_filter_or_exclude(args, kwargs)
+            return super(MultilingualQuerySet, self)._filter_or_exclude(negate, *args, **kwargs)
 
     def _get_original_fields(self):
         source = (self.model._meta.concrete_fields if hasattr(self.model._meta, 'concrete_fields')
@@ -362,12 +372,12 @@ class MultilingualQuerySet(models.query.QuerySet):
             return super(MultilingualQuerySet, self).create(**kwargs)
 
     # This method was not present in django-linguo
-    def get_or_create(self, **kwargs):
+    def get_or_create(self, *args, **kwargs):
         """
         Allows to override population mode with a ``populate`` method.
         """
         with auto_populate(self._populate_mode):
-            return super(MultilingualQuerySet, self).get_or_create(**kwargs)
+            return super(MultilingualQuerySet, self).get_or_create(*args, **kwargs)
 
     # This method was not present in django-linguo
     def defer(self, *fields):
@@ -384,23 +394,23 @@ class MultilingualQuerySet(models.query.QuerySet):
         return super(MultilingualQuerySet, self).values(*fields)
 
     def _values(self, *original, **kwargs):
-        if not kwargs.get('prepare', False):
-            return super(MultilingualQuerySet, self)._values(*original)
+        if not kwargs.pop('prepare', False):
+            return super(MultilingualQuerySet, self)._values(*original, **kwargs)
         new_fields, translation_fields = append_fallback(self.model, original)
-        clone = super(MultilingualQuerySet, self)._values(*list(new_fields))
+        clone = super(MultilingualQuerySet, self)._values(*list(new_fields), **kwargs)
         clone.original_fields = tuple(original)
         clone.translation_fields = translation_fields
         clone.fields_to_del = new_fields - set(original)
         return clone
 
     # This method was not present in django-linguo
-    def values(self, *fields):
+    def values(self, *fields, **expressions):
         if not self._rewrite:
-            return super(MultilingualQuerySet, self).values(*fields)
+            return super(MultilingualQuerySet, self).values(*fields, **expressions)
         if not fields:
             # Emulate original queryset behaviour: get all fields that are not translation fields
             fields = self._get_original_fields()
-        clone = self._values(*fields, prepare=True)
+        clone = self._values(*fields, prepare=True, **expressions)
         clone._iterable_class = FallbackValuesIterable
         return clone
 

